storage: remove placeholders immediately after Replica.handleRaftReady

petermattis opened the pull request:
This meshes better with per-Replica raft locking where
Store.processRaftMu will not be present to protect the placeholder
during application of a preemptive snapshot.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8956)

<!-- Reviewable:end -->

petermattis committed:
storage: remove placeholders immediately after Replica.handleRaftReady

This meshes better with per-Replica raft locking where
Store.processRaftMu will not be present to protect the placeholder
during application of a preemptive snapshot.

Always remove replica placeholders after preemptive snapshots

petermattis commented:
Cc @cockroachdb/stability

arjunravinarayan commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/store.go, line 1659 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQRuCHzzwRf-FuAXKxM:-KQRuCI-8vDLjZdyJxtE:b-8wl0pc) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc442d81a7931674ccca996836d6c4be9981f676/storage/store.go#L1659)):*
> ```Go
> }
> 
> func (s *Store) clearPlaceholder(rngID roachpb.RangeID) {
> ```

What's wrong with using `removePlaceholderLocked`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-:-KQRuMajQgnoN6woGvIw:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 1659 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQRuCHzzwRf-FuAXKxM:-KQRuy2Uo6v4PwP_XTV7:bpl1aui) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc442d81a7931674ccca996836d6c4be9981f676/storage/store.go#L1659)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> What's wrong with using `removePlaceholderLocked`?
</details>
Semantics there are different - that method returns an error if no placeholder is present.

I agree the two should be de-duplicated, though.

---

*[storage/store.go, line 2536 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQRvH8ap8BBA_6pR4Vw:-KQRvH8ap8BBA_6pR4Vx:b-rzxfb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc442d81a7931674ccca996836d6c4be9981f676/storage/store.go#L2536)):*
> ```Go
> 				}
> 				maybeWarnDuration(start, r, "handle raft ready")
> 				s.clearPlaceholder(r.RangeID)
> ```

should this be inside handleRaftReady?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 1659 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQRuCHzzwRf-FuAXKxM:-KQRzop8BYgPN6DFZB3K:b-ry3l0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc442d81a7931674ccca996836d6c4be9981f676/storage/store.go#L1659)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Semantics there are different - that method returns an error if no placeholder is present.
> 
> I agree the two should be de-duplicated, though.
</details>
I agree there is overlap, but it wasn't obvious to me (when writing this) how to de-duplicate these two methods. Any concrete suggestion?

---

*[storage/store.go, line 2536 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQRvH8ap8BBA_6pR4Vw:-KQS--kGDwnaNChk6fYf:b-fjierd) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc442d81a7931674ccca996836d6c4be9981f676/storage/store.go#L2536)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> should this be inside handleRaftReady?
</details>
Ok. Done. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/store.go, line 1659 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQRuCHzzwRf-FuAXKxM:-KQS9t9awnQIFMRxcvPJ:b-e3q13e) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc442d81a7931674ccca996836d6c4be9981f676/storage/store.go#L1659)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I agree there is overlap, but it wasn't obvious to me (when writing this) how to de-duplicate these two methods. Any concrete suggestion?
</details>
Well, `removePlaceHolderLocked` returns two possible errors:
- there is no placeholder with the given RangeID
- the state of `s.mu.replicaPlaceHolders` doesn't match that of `s.mu.replicasByKey

The former should just be returned and ignored at the new caller you've added. The latter should probably produce a `log.Fatal`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 1659 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQRuCHzzwRf-FuAXKxM:-KQSHW8UiG7fVH-FcpwS:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc442d81a7931674ccca996836d6c4be9981f676/storage/store.go#L1659)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Well, `removePlaceHolderLocked` returns two possible errors:
> - there is no placeholder with the given RangeID
> - the state of `s.mu.replicaPlaceHolders` doesn't match that of `s.mu.replicasByKey
> 
> The former should just be returned and ignored at the new caller you've added. The latter should probably produce a `log.Fatal`.
</details>
+1

---

*[storage/store.go, line 2544 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSF2g5X-aHUW1jxZxT:-KQSF2g5X-aHUW1jxZxU:bxcmt1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/a99169b2c1ff3a08330fc507ccf6469cc2067d68/storage/store.go#L2544)):*
> ```Go
> 			// the snapshot. Now that we have called handleRaftReady on all
> 			// replicas, clear all remaining placeholders.
> 			s.clearAllPlaceholders()
> ```

What I like about the old code is that there was nothing subtle about how long a placeholder would live - after each round of Raft processing, there would not be any left from before. Now this is more subtle - if a Raft command brings in a placeholder but then no Ready is produced, it is only the invocation of `handleRaftReady` being guaranteed by enqueuing a Raft update check (when handling the Raft message) that saves the day. That's brittle and needs to be documented and tested, though I suspect that bit of complexity is necessary.

Besides, handleRaftReady now always locks and unlocks the Store (those upcalls are also problematic in their own right) on every invocation, which can't be good for handling a lot of Replicas in parallel (as that would make Store.mu very contended). Perhaps that can be alleviated - the Store could supply the Replica with its placeholder (which could contain the unmarshaled snapshot, while we're at it) and so there wouldn't be a call to the Store lock unless there's actually a placeholder. Perhaps this doesn't work, but then the shortcoming should be pointed out explicitly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2544 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSF2g5X-aHUW1jxZxT:-KQSIuOK9Gfc22rk0sGg:b-2tkls6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a99169b2c1ff3a08330fc507ccf6469cc2067d68/storage/store.go#L2544)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What I like about the old code is that there was nothing subtle about how long a placeholder would live - after each round of Raft processing, there would not be any left from before. Now this is more subtle - if a Raft command brings in a placeholder but then no Ready is produced, it is only the invocation of `handleRaftReady` being guaranteed by enqueuing a Raft update check (when handling the Raft message) that saves the day. That's brittle and needs to be documented and tested, though I suspect that bit of complexity is necessary.
> 
> Besides, handleRaftReady now always locks and unlocks the Store (those upcalls are also problematic in their own right) on every invocation, which can't be good for handling a lot of Replicas in parallel (as that would make Store.mu very contended). Perhaps that can be alleviated - the Store could supply the Replica with its placeholder (which could contain the unmarshaled snapshot, while we're at it) and so there wouldn't be a call to the Store lock unless there's actually a placeholder. Perhaps this doesn't work, but then the shortcoming should be pointed out explicitly.
</details>
Yes, the old code was definitely more straightforward and yet it isn't tenable to keep that approach once we get rid of `processRaftMu`. 

Let me think about what can be done to avoid locking/unlocking `Store.mu` when there is no placeholder (the common case).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Reviewed 1 of 1 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1564 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSIg4f811qXg0gHFkd:-KQSIg4f811qXg0gHFke:bpbysh9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e25cc34310aa34c383e4ddef599760c31e6b224/storage/replica.go#L1564)):*
> ```Go
> 
> func (r *Replica) handleRaftReady() error {
> 	defer r.store.clearPlaceholder(r.RangeID)
> ```

Is it possible to refactor things so that we only only do this conditionally (if there is a placeholder for this range)? That would remove this contention on `r.store.mu`.
It seems to me that we could pass one bit of information to `handleRaftReady` that there is a placeholder that needs to be cleared, which would prevent the need to acquire the lock in most cases.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1564 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSIg4f811qXg0gHFkd:-KQSL7gaQ-vigo89Gbm2:bc7k2rl) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e25cc34310aa34c383e4ddef599760c31e6b224/storage/replica.go#L1564)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Is it possible to refactor things so that we only only do this conditionally (if there is a placeholder for this range)? That would remove this contention on `r.store.mu`.
> It seems to me that we could pass one bit of information to `handleRaftReady` that there is a placeholder that needs to be cleared, which would prevent the need to acquire the lock in most cases.
</details>
Seems legit.

---

*[storage/store.go, line 2544 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSF2g5X-aHUW1jxZxT:-KQSLBLKNgRT9M8V44NX:b-4jwwkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a99169b2c1ff3a08330fc507ccf6469cc2067d68/storage/store.go#L2544)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yes, the old code was definitely more straightforward and yet it isn't tenable to keep that approach once we get rid of `processRaftMu`. 
> 
> Let me think about what can be done to avoid locking/unlocking `Store.mu` when there is no placeholder (the common case).
</details>
@arjunravinarayan's suggestion may address this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Took another approach which is to only clear placeholders for uninitialized replicas. PTAL.

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1564 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSIg4f811qXg0gHFkd:-KQSUS-tUVXrFGblS3SD:b-8admxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e25cc34310aa34c383e4ddef599760c31e6b224/storage/replica.go#L1564)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Seems legit.
</details>
I've removed this call and now only call `clearPlaceholder` for uninitialized replicas.

---

*[storage/store.go, line 1659 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQRuCHzzwRf-FuAXKxM:-KQSWSvyn2hbk9jih0bv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc442d81a7931674ccca996836d6c4be9981f676/storage/store.go#L1659)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> +1
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-:-KQSXSHXdL15qo5aPVD9:b-jfgar0)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 1679 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSYIIa2qY3bgs1ci1Z:-KQSYIIa2qY3bgs1ci1_:b-7xpukf) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44d669d73937ee94c81cb357fed312729a78ce/storage/store.go#L1679)):*
> ```Go
> 		}
> 	}
> 	log.Fatalf(context.TODO(), "%s: placeholder %v was not in replicasByKey BTree", s, rngID)
> ```

nit: this is slightly clearer in an `else` to the above `if`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 1659 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSti1qYhuVqOgOWMvA:-KQSti1rUC7gTfm3eGQH:brarzdg) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44d669d73937ee94c81cb357fed312729a78ce/storage/store.go#L1659)):*
> ```Go
> }
> 
> func (s *Store) removePlaceholder(rngID roachpb.RangeID) bool {
> ```

Comment.

---

*[storage/store.go, line 2364 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQStoLvDyOsoe7Fug0T:-KQStoLvDyOsoe7Fug0U:b-35oeu6) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44d669d73937ee94c81cb357fed312729a78ce/storage/store.go#L2364)):*
> ```Go
> 				if !s.removePlaceholderLocked(req.RangeID) {
> 					return roachpb.NewError(
> 						errors.Errorf("%s: could not remove placeholder after preemptive snapshot", r))
> ```

Are we still consistent if this happens? ISTM that this should be Fatal.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2544 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSF2g5X-aHUW1jxZxT:-KQSuRVl17qZPJiF5fzX:b46iaar) ([raw file](https://github.com/cockroachdb/cockroach/blob/a99169b2c1ff3a08330fc507ccf6469cc2067d68/storage/store.go#L2544)):*
> What I like about the old code is that there was nothing subtle about how long a placeholder would live - after each round of Raft processing, there would not be any left from before. Now this is more subtle - if a Raft command brings in a placeholder but then no Ready is produced, it is only the invocation of handleRaftReady being guaranteed by enqueuing a Raft update check (when handling the Raft message) that saves the day. That's brittle and needs to be documented and tested, though I suspect that bit of complexity is necessary.

This part is still open.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 2544 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSF2g5X-aHUW1jxZxT:-KQSz4uvmgY7t28eCvOk:b-aei6wt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a99169b2c1ff3a08330fc507ccf6469cc2067d68/storage/store.go#L2544)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> > What I like about the old code is that there was nothing subtle about how long a placeholder would live - after each round of Raft processing, there would not be any left from before. Now this is more subtle - if a Raft command brings in a placeholder but then no Ready is produced, it is only the invocation of handleRaftReady being guaranteed by enqueuing a Raft update check (when handling the Raft message) that saves the day. That's brittle and needs to be documented and tested, though I suspect that bit of complexity is necessary.
> 
> This part is still open.
</details>
I've added a comment when the placeholder is added about the two possible removal paths. In doing so I noticed that we don't seem to remove the placeholder correctly for failed preemptive snapshots. I've added a `TODO(peter)`. 

Any thoughts on how to test this? I would have expected current tests to cover this code, but disabling the placeholder removal in `Store.processRaft` does not result in any failures.

---

*[storage/store.go, line 1659 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSti1qYhuVqOgOWMvA:-KQSxwWUWYJ-7eNWDCVL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44d669d73937ee94c81cb357fed312729a78ce/storage/store.go#L1659)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Comment.
</details>
Done.

---

*[storage/store.go, line 1679 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSYIIa2qY3bgs1ci1Z:-KQSxxQqbnjmwhQ9sMz3:b-ibm9v8) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44d669d73937ee94c81cb357fed312729a78ce/storage/store.go#L1679)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: this is slightly clearer in an `else` to the above `if`
</details>
I've refactored this slightly. PTAL.

---

*[storage/store.go, line 2364 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQStoLvDyOsoe7Fug0T:-KQSy25Q2dnZfXqVBOKm:b-rwt060) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44d669d73937ee94c81cb357fed312729a78ce/storage/store.go#L2364)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Are we still consistent if this happens? ISTM that this should be Fatal.
</details>
It is very unexpected for this to happen, though I'm not sure if we're inconsistent. I'm fine with changing it to a `log.Fatal`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 3 files at r3, 1 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 1659 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQTYlKyfj-UTMt4Rwfj:-KQTYlKzTeR10zyWKkc6:b-ljz4xd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e401369e6421136a5af21ba912c733861a181d0/storage/store.go#L1659)):*
> ```Go
> }
> 
> // removePlaceHolder removes a placeholder for the specified range if it
> ```

s/H/h/

---

*[storage/store.go, line 1688 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQTZ-IO3tksRd-7_3Mg:-KQTZ-IO3tksRd-7_3Mh:b-sdzbwf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e401369e6421136a5af21ba912c733861a181d0/storage/store.go#L1688)):*
> ```Go
> func (s *Store) addReplicaToRangeMapLocked(rng *Replica) error {
> 	if _, ok := s.mu.replicas[rng.RangeID]; ok {
> 		return errors.Errorf("%s: replica already existings", rng)
> ```

s/ing//

---

*[storage/store.go, line 2257 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQTZERgMv3KbNygR7li:-KQTZERgMv3KbNygR7lj:b-vkzyqm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e401369e6421136a5af21ba912c733861a181d0/storage/store.go#L2257)):*
> ```Go
> 	// Lazily create the replica.
> 	r, err := s.getOrCreateReplica(req.RangeID, req.ToReplica.ReplicaID, req.FromReplica)
> 	// TODO(bdarnell): is it safe to release the store lock here?
> ```

This comment is now confusing since the lock of Store.mu is hidden away in getOrCreateReplica.

---

*[storage/store.go, line 2646 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQTZOQd5bWJofuWoPLC:-KQTZOQd5bWJofuWoPLD:bv295ou) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e401369e6421136a5af21ba912c733861a181d0/storage/store.go#L2646)):*
> ```Go
> 
> // getOrCreateReplica returns a replica for the given RangeID, creating an
> // uninitialized replica if necessary. The caller must hold the store's lock.
> ```

s/must/must not/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-:-KQT_MAvELIMsGgK7wny:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 1679 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSYIIa2qY3bgs1ci1Z:-KQTYBvONTlj7G6YAVsy:b-uy9da7) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44d669d73937ee94c81cb357fed312729a78ce/storage/store.go#L1679)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I've refactored this slightly. PTAL.
</details>
still seems like it'd be clearer in an else

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 1679 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSYIIa2qY3bgs1ci1Z:-KQVpYswAEaSvArySkTD:b-z083di) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44d669d73937ee94c81cb357fed312729a78ce/storage/store.go#L1679)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> still seems like it'd be clearer in an else
</details>
Leaving as-is. Seems fine to me.

---

*[storage/store.go, line 1659 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQTYlKyfj-UTMt4Rwfj:-KQVpWs-EakoIdkmJLMe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e401369e6421136a5af21ba912c733861a181d0/storage/store.go#L1659)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/H/h/
</details>
Done.

---

*[storage/store.go, line 1688 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQTZ-IO3tksRd-7_3Mg:-KQVpb9ACysIvnu_9RbC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e401369e6421136a5af21ba912c733861a181d0/storage/store.go#L1688)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/ing//
</details>
Done.

---

*[storage/store.go, line 2257 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQTZERgMv3KbNygR7li:-KQVpcoSODnTnDwBgJnn:b-27sd1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e401369e6421136a5af21ba912c733861a181d0/storage/store.go#L2257)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This comment is now confusing since the lock of Store.mu is hidden away in getOrCreateReplica.
</details>
Yeah, I wanted to ask you what this is in reference to. The comment has been here a long time. Should I just nuke it?

---

*[storage/store.go, line 2646 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQTZOQd5bWJofuWoPLC:-KQVpzqpy8ogbAsnI36z:bnbd0b3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e401369e6421136a5af21ba912c733861a181d0/storage/store.go#L2646)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/must/must not/
</details>
Done. Also, reformatted the excessively long line below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2544 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSF2g5X-aHUW1jxZxT:-KQWKZtCNqHI5L4mth78:b-5kb0iu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a99169b2c1ff3a08330fc507ccf6469cc2067d68/storage/store.go#L2544)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I've added a comment when the placeholder is added about the two possible removal paths. In doing so I noticed that we don't seem to remove the placeholder correctly for failed preemptive snapshots. I've added a `TODO(peter)`. 
> 
> Any thoughts on how to test this? I would have expected current tests to cover this code, but disabling the placeholder removal in `Store.processRaft` does not result in any failures.
</details>
Ping. @arjunravinarayan, given that you were the original author of this code, do you have any thoughts on testing it more thoroughly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tamird commented:
Should this be just one commit? I'm confused trying to review this as two.

---

Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[storage/store.go, line 1679 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSYIIa2qY3bgs1ci1Z:-KQWRC7JKJsiC5brFv0R:b-odpsa2) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44d669d73937ee94c81cb357fed312729a78ce/storage/store.go#L1679)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Leaving as-is. Seems fine to me.
</details>
```
switch exRng := s.mu.replicasByKey.Delete(rng).(type); exRng {
case *ReplicaPlaceholder:
  delete(s.mu.replicaPlaceholders, rngID)
  return true
case nil:
  log.Fatalf(context.TODO(), "%s range=%d: placeholder not found", s, rngID)
default:
  log.Fatalf(context.TODO(), "%s range=%d: expected placeholder, got %T", s, rngID, exRng)
}
return false // appease the compiler
```

Feel free to ignore.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-:-KQWPxfv2cT01lqAGB7i:bwif20v)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Squashed.

---

Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 1679 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSYIIa2qY3bgs1ci1Z:-KQWWo0tCR3lvDkqRDEU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44d669d73937ee94c81cb357fed312729a78ce/storage/store.go#L1679)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ```
> switch exRng := s.mu.replicasByKey.Delete(rng).(type); exRng {
> case *ReplicaPlaceholder:
>   delete(s.mu.replicaPlaceholders, rngID)
>   return true
> case nil:
>   log.Fatalf(context.TODO(), "%s range=%d: placeholder not found", s, rngID)
> default:
>   log.Fatalf(context.TODO(), "%s range=%d: expected placeholder, got %T", s, rngID, exRng)
> }
> return false // appease the compiler
> ```
> 
> Feel free to ignore.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-:-KQWWkMR_eQORyotVM3u:b-un4k39)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 1679 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSYIIa2qY3bgs1ci1Z:-KQWXlpRfTFwBN5RjR6l:b-mx3knv) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44d669d73937ee94c81cb357fed312729a78ce/storage/store.go#L1679)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Done.
</details>
Really done (with the compile error fixed).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 1679 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSYIIa2qY3bgs1ci1Z:-KQWdHgPeSvmQ3IEBrda:b-5pzhiu) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44d669d73937ee94c81cb357fed312729a78ce/storage/store.go#L1679)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Really done (with the compile error fixed).
</details>
Ah, I guess I meant to write `switch exRng := s.mu.replicasByKey.Delete(rng).(type) {`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-:-KQWdMVYPUZvldGv7cPU:bnfp4nl)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
Reviewed 1 of 3 files at r3, 1 of 2 files at r4, 1 of 1 files at r5, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2544 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSF2g5X-aHUW1jxZxT:-KQWM7QII_9-MUCjfOfw:b-vrpbtm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a99169b2c1ff3a08330fc507ccf6469cc2067d68/storage/store.go#L2544)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Ping. @arjunravinarayan, given that you were the original author of this code, do you have any thoughts on testing it more thoroughly.
</details>
Tricky. `ChangeReplicas` does a lot of checking, so it's hard to exercise this code path using a pair of dueling preemptive snapshots - it will get caught by other checks (including `hasOverlappingKeyRange`). Hence why my test was very basic and just testing the placeholder addition and deletion directly. I was sort of relying on the `clearAllPlaceholders` crutch pretty heavily in order to guarantee the short-lifetime invariant.

To test the code path, we need to send a snapshot which will clear all the checks in `ChangeReplicas` and `canApplySnapshot`, but will be rejected by raft and not be present in the `Ready` that arrives at`handleRaftReady`. We then need to send a second snapshot that will be accepted. This will verify that the placeholder created for the first snapshot is cleared in time for the second snapshot. I don't know how to achieve this - It seems to me that there's no way to construct a preemptive snapshot that gets rejected by Raft. But perhaps it looks something along the lines of this: (@bdarnell do you have any idea how to make this concrete?)

1. Create a `multiTestContext` with 3 stores.
2. Create a range on stores 1, replicate to store 2.
3. Manually replicate to store 3 - Don't directly use `multiTestContext.replicateRange`, but do what it does in manually constructing a call to `ChangeReplicas`. Somehow (how?) mess up the snapshot so that store 3 will reject it only at the Raft level. `ChangeReplicas` looks pretty robust to me - it fetches all the information from the snapshot, and anything you give it goes through the local replica, so it's not possible to construct something that has malformed IDs/other metadata and still have the snapshot be sent.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 1 of 2 files at r4, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1619 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQWoh7jaikUEPrnmoYw:-KQWoh7jaikUEPrnmoYx:bqlaphq) ([raw file](https://github.com/cockroachdb/cockroach/blob/529296a6f2bef7b1862d93dcbfbea517d621cf76/storage/replica.go#L1619)):*
> ```Go
> 			defer r.store.mu.Unlock()
> 
> 			r.store.removePlaceholderLocked(r.RangeID)
> ```

I don't think this addresses the main point in my earlier comments. If the Replica doesn't produce a snapshot, when do we remove the placeholder? We return early in `handleRaftReady`.

---

*[storage/store.go, line 2416 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQWpl7obA8DtZpSJd9H:-KQWpl7pNBtKiPyxcRyM:b-nkiaa3) ([raw file](https://github.com/cockroachdb/cockroach/blob/529296a6f2bef7b1862d93dcbfbea517d621cf76/storage/store.go#L2416)):*
> ```Go
> 	}
> 
> 	s.enqueueRaftUpdateCheck(req.RangeID)
> ```

Makes me mildly uneasy that this is so far down even though we crucially rely on this being called much higher up (for example, if you create a placeholder but then some random thing below gives an error and early return, we have, at least for that cycle, leaked a placeholder - can be bad if no activity is on that range otherwise since it blocks incoming snapshots)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1619 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQWoh7jaikUEPrnmoYw:-KQWvYsthPfTXWJxw2iK:bl4hpwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/529296a6f2bef7b1862d93dcbfbea517d621cf76/storage/replica.go#L1619)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I don't think this addresses the main point in my earlier comments. If the Replica doesn't produce a snapshot, when do we remove the placeholder? We return early in `handleRaftReady`.
</details>
Placeholders are only added for uninitialized replicas. We always remove the placeholder after calling `handleRaftReady` on an uninitialized replica in `Store.processRaft`. Am I misunderstanding your concern. (The change here was cosmetic).

---

*[storage/store.go, line 1679 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSYIIa2qY3bgs1ci1Z:-KQWvLqv1Q5YciBGPE4s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be44d669d73937ee94c81cb357fed312729a78ce/storage/store.go#L1679)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Ah, I guess I meant to write `switch exRng := s.mu.replicasByKey.Delete(rng).(type) {`
</details>
Done.

---

*[storage/store.go, line 2416 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQWpl7obA8DtZpSJd9H:-KQWvoIcmf2k_97033TS:bmxk28d) ([raw file](https://github.com/cockroachdb/cockroach/blob/529296a6f2bef7b1862d93dcbfbea517d621cf76/storage/store.go#L2416)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Makes me mildly uneasy that this is so far down even though we crucially rely on this being called much higher up (for example, if you create a placeholder but then some random thing below gives an error and early return, we have, at least for that cycle, leaked a placeholder - can be bad if no activity is on that range otherwise since it blocks incoming snapshots)
</details>
Yeah, I think there are still holes here. I'm going to refactor this so that the `defer` to remove the placeholder is done in a different place. I'll ping this again when that's ready.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
*[storage/replica.go, line 1619 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQWoh7jaikUEPrnmoYw:-KQWvvJKAzX_N6bVvTX-:b-yn793f) ([raw file](https://github.com/cockroachdb/cockroach/blob/529296a6f2bef7b1862d93dcbfbea517d621cf76/storage/replica.go#L1619)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Placeholders are only added for uninitialized replicas. We always remove the placeholder after calling `handleRaftReady` on an uninitialized replica in `Store.processRaft`. Am I misunderstanding your concern. (The change here was cosmetic).
</details>
I'm with @petermattis here - it looks fine to me. `store.go:2556` guarantees that `removePlaceholderLocked` is always called after every invocation of `handleRaftReady()` on an uninitialized replica. And there are no placeholders for initialized replicas.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1619 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQWoh7jaikUEPrnmoYw:-KQWyQ14jjOyp_igoRPC:bog5381) ([raw file](https://github.com/cockroachdb/cockroach/blob/529296a6f2bef7b1862d93dcbfbea517d621cf76/storage/replica.go#L1619)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> I'm with @petermattis here - it looks fine to me. `store.go:2556` guarantees that `removePlaceholderLocked` is always called after every invocation of `handleRaftReady()` on an uninitialized replica. And there are no placeholders for initialized replicas.
</details>
You're right, I didn't have that bit of the diff paged in. Apologies.

---

*[storage/store.go, line 2556 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQWyH-1EQ0KRbE_DWlj:-KQWyH-1EQ0KRbE_DWlk:bbbffsf) ([raw file](https://github.com/cockroachdb/cockroach/blob/529296a6f2bef7b1862d93dcbfbea517d621cf76/storage/store.go#L2556)):*
> ```Go
> 				// definition, an initialized replica will be present in the
> 				// replicasByKey map.
> 				s.removePlaceholder(r.RangeID)
> ```

Add to the comment that while the replica will usually consume its placeholder itself, that isn't guaranteed and so this invocation here is crucial (someone might get smart ideas and remove this one).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 3 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 2268 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQX6EATk8S2XtBjFDPH:-KQX6EATk8S2XtBjFDPI:b6yzkcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0358e27ab7c2686e33a7f25babede6a80692965/storage/store.go#L2268)):*
> ```Go
> 				}
> 			}()
> 		}
> ```

@tschottdorf Does this address your concern about about placeholder removal? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[storage/store.go, line 2544 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQSF2g5X-aHUW1jxZxT:-KQXBhU8_SSYZp9nWr16:b-fu2dqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a99169b2c1ff3a08330fc507ccf6469cc2067d68/storage/store.go#L2544)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> Tricky. `ChangeReplicas` does a lot of checking, so it's hard to exercise this code path using a pair of dueling preemptive snapshots - it will get caught by other checks (including `hasOverlappingKeyRange`). Hence why my test was very basic and just testing the placeholder addition and deletion directly. I was sort of relying on the `clearAllPlaceholders` crutch pretty heavily in order to guarantee the short-lifetime invariant.
> 
> To test the code path, we need to send a snapshot which will clear all the checks in `ChangeReplicas` and `canApplySnapshot`, but will be rejected by raft and not be present in the `Ready` that arrives at`handleRaftReady`. We then need to send a second snapshot that will be accepted. This will verify that the placeholder created for the first snapshot is cleared in time for the second snapshot. I don't know how to achieve this - It seems to me that there's no way to construct a preemptive snapshot that gets rejected by Raft. But perhaps it looks something along the lines of this: (@bdarnell do you have any idea how to make this concrete?)
> 
> 1. Create a `multiTestContext` with 3 stores.
> 2. Create a range on stores 1, replicate to store 2.
> 3. Manually replicate to store 3 - Don't directly use `multiTestContext.replicateRange`, but do what it does in manually constructing a call to `ChangeReplicas`. Somehow (how?) mess up the snapshot so that store 3 will reject it only at the Raft level. `ChangeReplicas` looks pretty robust to me - it fetches all the information from the snapshot, and anything you give it goes through the local replica, so it's not possible to construct something that has malformed IDs/other metadata and still have the snapshot be sent.
</details>
You can probably also manually craft a snapshot that raft will judge to be "old" by messing with its metadata.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 2268 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQX6EATk8S2XtBjFDPH:-KQXCa5EQAUYBp-tAoQR:b-af1832) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0358e27ab7c2686e33a7f25babede6a80692965/storage/store.go#L2268)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> @tschottdorf Does this address your concern about about placeholder removal? 
</details>
Yep, thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-:-KQXCijZpOA0kp2F24wJ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I've added a test that we remove the placeholder on certain previously unhandled error conditions from `Store.HandleRaftRequest`. I'm still trying to figure out how to create a snapshot that will create a placeholder but then be dropped by raft.

petermattis commented:
Added another test which verifies we remove placeholders for dropped snapshots. I've verified that the `removePlaceholder` call in `Store.processRaft` is required, though the test itself seems a bit fragile (i.e. it could pass if the placeholder was never added). I'm not sure how to fix that.

tamird commented:
Reviewed 1 of 2 files at r10, 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[storage/store_test.go, line 2294 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQXXe1YzuGYitwYUHCb:-KQXXe1YzuGYitwYUHCc:b-oct2ii) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc9536bc53f5cbc319393344d1945b4a05a3ab33/storage/store_test.go#L2294)):*
> ```Go
> 		},
> 	}
> 	if err := s.HandleRaftRequest(ctx, req); err == nil {
> ```

prefer `testutils.IsError`

---

*[storage/store_test.go, line 2297 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQXXhfhUopkEGqDuEWD:-KQXXhfhUopkEGqDuEWE:b2ribkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc9536bc53f5cbc319393344d1945b4a05a3ab33/storage/store_test.go#L2297)):*
> ```Go
> 		t.Fatalf("expected error, but found success")
> 	} else {
> 		fmt.Printf("%s\n", err)
> ```

remove (instead log unexpected errors)

---

*[storage/store_test.go, line 2301 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQXXn6nUcgJMNhFEa0Z:-KQXXn6nUcgJMNhFEa0_:b-i1w415) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc9536bc53f5cbc319393344d1945b4a05a3ab33/storage/store_test.go#L2301)):*
> ```Go
> 
> 	s.mu.Lock()
> 	numPlaceholders := len(s.mu.replicaPlaceholders)
> ```

hm. what is this test actually testing? in other words, how do i know that a placeholder was created in the first place?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Agree with @tamird that while the tests probably give good coverage, they're likely hard to reason about and could rot without failing. Perhaps it's enough to add the number of placeholders discarded as a metric (that is the number that we usually want to see close to zero in production as well).

---

Reviewed 1 of 2 files at r10, 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/store_test.go, line 2294 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQXXe1YzuGYitwYUHCb:-KQXgl8l3JTJ9llml2Us:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc9536bc53f5cbc319393344d1945b4a05a3ab33/storage/store_test.go#L2294)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> prefer `testutils.IsError`
</details>
+1

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-:-KQXhmFkdVF_898mvUfL:b34kxh9)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 2556 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQWyH-1EQ0KRbE_DWlj:-KQXECTQj4kFu1ECVwvh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/529296a6f2bef7b1862d93dcbfbea517d621cf76/storage/store.go#L2556)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Add to the comment that while the replica will usually consume its placeholder itself, that isn't guaranteed and so this invocation here is crucial (someone might get smart ideas and remove this one).
</details>
Done.

---

*[storage/store_test.go, line 2294 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQXXe1YzuGYitwYUHCb:-KQY6YOCy3TPa44darTY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc9536bc53f5cbc319393344d1945b4a05a3ab33/storage/store_test.go#L2294)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> +1
</details>
Done.

---

*[storage/store_test.go, line 2297 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQXXhfhUopkEGqDuEWD:-KQY6YoMSaAZIJpSz_TW:bhedpzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc9536bc53f5cbc319393344d1945b4a05a3ab33/storage/store_test.go#L2297)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> remove (instead log unexpected errors)
</details>
Removed (accidentally left in).

---

*[storage/store_test.go, line 2301 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQXXn6nUcgJMNhFEa0Z:-KQY6_uypRvogZ_kNbcd:bocijw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/bc9536bc53f5cbc319393344d1945b4a05a3ab33/storage/store_test.go#L2301)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> hm. what is this test actually testing? in other words, how do i know that a placeholder was created in the first place?
</details>
See my comment on the PR. I've added counts on the number of removed, filled and dropped placeholders and checks that the corresponding value is as expected here and in the other test. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Latest changes reduce the test fragility so that the tests will now fail if we never created a placeholder. I added simple counts, not metrics since I'm not sure if we want/need metrics for the placeholder stats. We can change that in the future if needed.

---

Review status: 0 of 3 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-:-KQY6qMo0z0lablevQK2:bjz68l8)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 3 of 3 files at r12.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 2257 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQTZERgMv3KbNygR7li:-KQZShSw2wExhKw7ixyq:bc1tigd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e401369e6421136a5af21ba912c733861a181d0/storage/store.go#L2257)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yeah, I wanted to ask you what this is in reference to. The comment has been here a long time. Should I just nuke it?
</details>
Yeah, delete it. I think it predates processRaftMu and is about whether it's safe to create a replica here, release all locks, then locking the replica and calling raftGroup.Step. 

---

*[storage/store.go, line 2270 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQZT9PgtRj3f2hde7pZ:-KQZT9PgtRj3f2hde7p_:b-uc7cz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/68b9632e7e417f6e3d33064a790f3600389b78a6/storage/store.go#L2270)):*
> ```Go
> 			// If we added a placeholder remove it before we return unless some other
> 			// part of the code takes ownership of the removal (indicated by setting
> 			// removePlaceholder to true).
> ```

s/true/false/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r12.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 423 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQ_Ka0gYg9VmE-6WvOC:-KQ_Ka0hrdr3tO0-FK9Z:b-xzuvdj) ([raw file](https://github.com/cockroachdb/cockroach/blob/68b9632e7e417f6e3d33064a790f3600389b78a6/storage/store.go#L423)):*
> ```Go
> 		// Number of placeholders successfully filled by a snapshot.
> 		filledPlaceholders int32
> 		// Number of placeholders removed a snapshot was dropped by raft.
> ```

broken sentence

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 2 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[storage/store.go, line 2257 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQTZERgMv3KbNygR7li:-KQ_hA0ciuzGOK0f5H44:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e401369e6421136a5af21ba912c733861a181d0/storage/store.go#L2257)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Yeah, delete it. I think it predates processRaftMu and is about whether it's safe to create a replica here, release all locks, then locking the replica and calling raftGroup.Step. 
</details>
Done.

---

*[storage/store.go, line 423 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQ_Ka0gYg9VmE-6WvOC:-KQ_h17UeKGavkGNcAtO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/68b9632e7e417f6e3d33064a790f3600389b78a6/storage/store.go#L423)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> broken sentence
</details>
Done.

---

*[storage/store.go, line 2270 \[r12\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956#-KQZT9PgtRj3f2hde7pZ:-KQ_h4S08CwpNRxTcrAJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/68b9632e7e417f6e3d33064a790f3600389b78a6/storage/store.go#L2270)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/true/false/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r13.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8956)*
<!-- Sent from Reviewable.io -->
